src/VisualStudio/Core/Def/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -229,108 +247,121 @@ private WatcherOperation(Kind kind, Context.RegularWatchedFile token)
-            public bool TryCombineWith(in WatcherOperation other, out WatcherOperation combined)
+            /// <param name="operations">The collection containing the operations to combine.</param>
+            /// <param name="start">Start index (inclusive) of operations to combine.</param>
+            /// <param name="end">End index (inclusive) of operations to combine.</param>
+            public static WatcherOperation CombineRange(ImmutableSegmentedList<WatcherOperation> operations, int start, int end)
             {
-                if (other._kind == Kind.None)
-                {
-                    combined = this;
-                    return true;
-                }
-                else if (_kind == Kind.None)
-                {
-                    combined = other;
-                    return true;
-                }
-
-                switch (_kind)
-                {
-                    case Kind.WatchDirectory:
-                    case Kind.WatchFile:
-                        // Watching operations cannot be combined
-                        break;
-
-                    case Kind.UnwatchFile when other._kind == Kind.UnwatchFile:
-                        combined = UnwatchFiles(ImmutableList.Create(_token, other._token));
-                        return true;
-
-                    case Kind.UnwatchFile when other._kind == Kind.UnwatchFiles:
-                        combined = UnwatchFiles(other._tokens.ToImmutableList().Insert(0, _token));
-                        return true;
+                var firstOp = operations[start];
+                if (start == end)
+                    return firstOp;
 
-                    case Kind.UnwatchDirectories when other._kind == Kind.UnwatchDirectories:
-                        var cookies = new List<uint>(_cookies);
-                        cookies.AddRange(other._cookies);
-                        combined = UnwatchDirectories(cookies);
-                        return true;
+                using var _1 = ArrayBuilder<Context.RegularWatchedFile>.GetInstance(out var tokensBuilder);
+                using var _2 = ArrayBuilder<string>.GetInstance(out var fileNamesBuilder);
+                using var _3 = ArrayBuilder<uint>.GetInstance(out var cookiesBuilder);
 
-                    case Kind.UnwatchFiles when other._kind == Kind.UnwatchFile:
-                        combined = UnwatchFiles(_tokens.ToImmutableList().Add(other._token));
-                        return true;
-
-                    case Kind.UnwatchFiles when other._kind == Kind.UnwatchFiles:
-                        combined = UnwatchFiles(_tokens.ToImmutableList().AddRange(other._tokens));
-                        return true;
+                for (; start <= end; start++)
+                {
+                    var op = operations[start];
 
-                    default:
-                        break;
+                    switch (op._kind)
+                    {
```

Comment:
What worries me most with this is whether we'd accidentally forget to include one of the fields in one of the branches, meaning a combining might drop something. Is there any reason we can't do something like just checking _token for whether there is one, or just adding _tokens (which will be empty and probably a no-op)?

Reply:
Wasn't clear on the ask here, are you still wanting this change?

Reply:
I don't think I'd want to merge these codepaths. One concern is that some of these fields could be null (eg, op.__cookies can be null). It's not a lot of code, and I think it makes it clearer if they are separated.